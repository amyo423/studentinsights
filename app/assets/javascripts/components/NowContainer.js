import React from 'react';

// A container for testing that allows passing in a `nowFn`
// that will be shared with child components through context.
export default class NowContainer extends React.Component {
  getChildContext() {
    const {nowFn} = this.props;
    return {
      nowFn() { return nowFn(); }
    };
  }

  render() {
    const {children} = this.props;

    return children;
  }
}
NowContainer.childContextTypes = {
  nowFn: React.PropTypes.func
};
NowContainer.propTypes = {
  children: React.PropTypes.node.isRequired,
  nowFn: React.PropTypes.func.isRequired
};