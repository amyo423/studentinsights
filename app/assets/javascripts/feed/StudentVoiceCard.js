import PropTypes from 'prop-types';
import React from 'react';
import _ from 'lodash';
import hash from 'object-hash';
import Card from '../components/Card';


// Render a card in the feed showing there are new student voice surveys in,
// and allow clicking to see list of students and then jump to their profiles.
export default class StudentVoiceCard extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isExpanded: false,
      shuffleSeed: props.shuffleSeed || _.random(0, 32)
    };

    this.onExpandClicked = this.onExpandClicked.bind(this);
  }

  onExpandClicked(e) {
    e.preventDefault();
    this.setState({isExpanded: true});
  }

  render() {
    const {style = {}} = this.props;
    return (
      <Card className="StudentVoiceCard" style={style}>
        {this.renderStudentsAndCount()}        
      </Card>
    );
  }

  renderStudentsAndCount() {
    const {studentVoiceCardJson} = this.props;
    const {students} = studentVoiceCardJson;
    const {isExpanded, shuffleSeed} = this.state;
    const emoji = <span style={{position: 'relative', top: 1}}>ðŸ’¬</span>;

    // If the list is <=3 students, show them all inline with links.
    // If it's more, pick two random to show and then +n more, which expands.
    if (students.length === 0) {
      return null; // shouldn't happen, but guard
    } else if (students.length === 1) {
      return <div>{emoji} {this.renderStudentLink(students[0])} shared a new student voice survey.</div>;
    } else if (students.length === 2) {
      return <div>{emoji} {this.renderStudentLink(students[0])} and {this.renderStudentLink(students[1])} shared new student voice surveys.</div>;
    } else if (students.length === 3) {
      return <div>{emoji} {this.renderStudentLink(students[0])}, {this.renderStudentLink(students[1])} and {this.renderStudentLink(students[2])} shared new student voice surveys.</div>;
    }

    const shuffledStudents = _.sortBy(students, student => hash({...student, shuffleSeed}));
    if (!isExpanded) {
      return <div>{emoji} {this.renderStudentLink(shuffledStudents[0])}, {this.renderStudentLink(shuffledStudents[1])} and {this.renderExpandableListLink(shuffledStudents.slice(2))} shared new student voice surveys.</div>;
    } else {
      return <div>{emoji} {this.renderStudentLink(shuffledStudents[0])}, {this.renderStudentLink(shuffledStudents[1])} and {this.renderMoreStudentsText(shuffledStudents.slice(2))} shared new student voice surveys.{this.renderExpandedList(shuffledStudents.slice(2))}</div>;
    }
  }

  renderMoreStudentsText(restOfStudents) {
    return `${restOfStudents.length} more students`;
  }

  renderExpandableListLink(restOfStudents) {
    return <a href="#" onClick={this.onExpandClicked}>{this.renderMoreStudentsText(restOfStudents)}</a>;
  }

  renderExpandedList(restOfStudents) {
    return (
      <div style={{paddingTop: 10, paddingLeft: 10}}>
        {restOfStudents.map(student => (
          <div key={student.id} style={{padding: 5}}>{this.renderStudentLink(student)}</div>
        ))}
      </div>
    );
  }

  renderStudentLink(student) {
    return <a style={{fontWeight: 'bold'}} href={`/students/${student.id}`}>{student.first_name} {student.last_name}</a>;
  }
}
StudentVoiceCard.propTypes = {
  studentVoiceCardJson: PropTypes.shape({
    latest_form_timestamp: PropTypes.string.isRequired,
    imported_forms_for_date_count: PropTypes.number.isRequired,
    students: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      first_name: PropTypes.string.isRequired,
      last_name: PropTypes.string.isRequired
    })).isRequired
  }).isRequired,
  style: PropTypes.object,
  shuffleSeed: PropTypes.number
};
