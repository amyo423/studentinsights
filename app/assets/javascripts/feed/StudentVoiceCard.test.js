import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import StudentVoiceCard from './StudentVoiceCard';
import {withDefaultNowContext} from '../testing/NowContainer';

function testStudents() {
  return [{
    id: 4,
    first_name: 'Luke',
    last_name: 'Skywalker'
  }, {
    id: 5,
    first_name: 'Mari',
    last_name: 'Kenobi'
  }, {
    id: 6,
    first_name: 'Kylo',
    last_name: 'Ren'
  }, {
    id: 7,
    first_name: 'Darth',
    last_name: 'Tater'
  }];
}

export function testProps(props = {}) {
  return {
    shuffleSeed: 42,
    studentVoiceCardJson: {
      latest_form_timestamp: '2018-03-11T11:03:00.000Z',
      imported_forms_for_date_count: 2,
      students: testStudents()
    },
    ...props
  };
}

export function propsWithNStudents(n) {
  const props = testProps();
  return {
    ...props,
    studentVoiceCardJson: {
      ...props.studentVoiceCardJson,
      students: testStudents().slice(0, n)
    }
  };
}


export function testEl(props) {
  return withDefaultNowContext(<StudentVoiceCard {...props} />);
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  ReactDOM.render(testEl(testProps()), el);
  expect($(el).text()).toContain('ðŸ’¬ Darth Tater, Kylo Ren and 2 more students shared new student voice surveys.');
  expect($(el).find('a').toArray().map(el => $(el).attr('href'))).toEqual([
    '/students/7',
    '/students/6',
    '#'
  ]);
});

describe('snapshots', () => {
  it('works with one student', () => {
    const tree = renderer.create(testEl(propsWithNStudents(1))).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('works with two students', () => {
    const tree = renderer.create(testEl(propsWithNStudents(2))).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('works with three students', () => {
    const tree = renderer.create(testEl(propsWithNStudents(3))).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('works with four students', () => {
    const tree = renderer.create(testEl(propsWithNStudents(4))).toJSON();
    expect(tree).toMatchSnapshot();
  });
});