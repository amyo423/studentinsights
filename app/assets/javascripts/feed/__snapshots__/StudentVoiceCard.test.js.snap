// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots works with four students 1`] = `
<div
  className="Card StudentVoiceCard"
  style={
    Object {
      "border": "1px solid #ccc",
      "borderRadius": 3,
      "padding": 10,
    }
  }
>
  <div>
    <span
      style={
        Object {
          "position": "relative",
          "top": 1,
        }
      }
    >
      ðŸ’¬
    </span>
     
    <a
      href="/students/7"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Darth
       
      Tater
    </a>
    , 
    <a
      href="/students/6"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Kylo
       
      Ren
    </a>
     and 
    <a
      href="#"
      onClick={[Function]}
    >
      2 more students
    </a>
     shared new student voice surveys.
  </div>
</div>
`;

exports[`snapshots works with one student 1`] = `
<div
  className="Card StudentVoiceCard"
  style={
    Object {
      "border": "1px solid #ccc",
      "borderRadius": 3,
      "padding": 10,
    }
  }
>
  <div>
    <span
      style={
        Object {
          "position": "relative",
          "top": 1,
        }
      }
    >
      ðŸ’¬
    </span>
     
    <a
      href="/students/4"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Luke
       
      Skywalker
    </a>
     shared a new student voice survey.
  </div>
</div>
`;

exports[`snapshots works with three students 1`] = `
<div
  className="Card StudentVoiceCard"
  style={
    Object {
      "border": "1px solid #ccc",
      "borderRadius": 3,
      "padding": 10,
    }
  }
>
  <div>
    <span
      style={
        Object {
          "position": "relative",
          "top": 1,
        }
      }
    >
      ðŸ’¬
    </span>
     
    <a
      href="/students/4"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Luke
       
      Skywalker
    </a>
    , 
    <a
      href="/students/5"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Mari
       
      Kenobi
    </a>
     and 
    <a
      href="/students/6"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Kylo
       
      Ren
    </a>
     shared new student voice surveys.
  </div>
</div>
`;

exports[`snapshots works with two students 1`] = `
<div
  className="Card StudentVoiceCard"
  style={
    Object {
      "border": "1px solid #ccc",
      "borderRadius": 3,
      "padding": 10,
    }
  }
>
  <div>
    <span
      style={
        Object {
          "position": "relative",
          "top": 1,
        }
      }
    >
      ðŸ’¬
    </span>
     
    <a
      href="/students/4"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Luke
       
      Skywalker
    </a>
     and 
    <a
      href="/students/5"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Mari
       
      Kenobi
    </a>
     shared new student voice surveys.
  </div>
</div>
`;
